unitary/alpha/quantum_world.py

```diff
@@ -689,26 +690,103 @@ def density_matrix(
-            cirq.von_neumann_entropy(density_matrix_1, validate=False)
-            + cirq.von_neumann_entropy(density_matrix_2, validate=False)
-            - cirq.von_neumann_entropy(density_matrix_12.reshape(4, 4), validate=False)
+        num_involved_objects = (
+            len(objects) if objects is not None else len(self.object_name_dict.values())
         )
 
+        if num_involved_objects < 2:
+            raise ValueError(
+                f"Could not calculate entanglement for {num_involved_objects} qubit. "
+                "At least 2 qubits are required."
+            )
+
+        involved_objects = (
+            objects if objects is not None else list(self.object_name_dict.values())
+        )
+
+        density_matrix = self.density_matrix(involved_objects)
+        reshaped_density_matrix = density_matrix.reshape((2, 2) * num_involved_objects)
+        result = 0.0
+        for comb in combinations(range(num_involved_objects), num_involved_objects - 1):
+            reshaped_partial_density_matrix = cirq.partial_trace(
+                reshaped_density_matrix, list(comb)
+            )
+            partial_density_matrix = reshaped_partial_density_matrix.reshape(
+                2 ** (num_involved_objects - 1), 2 ** (num_involved_objects - 1)
+            )
+            result += cirq.von_neumann_entropy(partial_density_matrix, validate=False)
+        result -= cirq.von_neumann_entropy(density_matrix, validate=False)
+        return result
+
+    def print_entanglement_table(self, count: int = 1000) -> None:
+        """Peek the current quantum world `count` times, and calculate pair-wise entanglement
+        (i.e. quantum mutual information) for each pair of quantum objects.
+        See https://en.wikipedia.org/wiki/Quantum_mutual_information for the formula. And print
+        the results out in a table.
+
+        Parameters:
+            count:      Number of measurements.
+        """
+        objects = list(self.object_name_dict.values())
+        num_qubits = len(objects)
+        if num_qubits < 2:
+            raise ValueError(
+                f"There is only {num_qubits} qubit in the quantum world. "
+                "At least 2 qubits are required to calculate entanglements."
+            )
+        # Peek the current world `count` times and get the results.
+        histogram = self.get_correlated_histogram(objects, count)
+
+        # Get an estimate of the state vector.
+        state_vector = np.array([0.0] * (2**num_qubits))
+        for key, val in histogram.items():
+            state_vector += self.__to_state_vector__(key) * np.sqrt(val * 1.0 / count)
+        density_matrix = np.outer(state_vector, state_vector)
+        reshaped_density_matrix = density_matrix.reshape((2, 2) * num_qubits)
+
+        entropy = [0.0] * num_qubits
+        entropy_pair = np.zeros((num_qubits, num_qubits))
+        entanglement = np.zeros((num_qubits, num_qubits))
+        for i in range(num_qubits - 1):
```

Comment:
I think this block is fairly tough to follow.  I would add comments below or in the function's docstring to explain the implementation details.